src/controller/CHIPDeviceController.h

```diff
@@ -714,108 +715,84 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
 
     CommissioneeDeviceProxy * FindCommissioneeDevice(const SessionHandle & session);
     CommissioneeDeviceProxy * FindCommissioneeDevice(NodeId id);
     void ReleaseCommissioneeDevice(CommissioneeDeviceProxy * device);
 
-    // Cluster callbacks for advancing commissioning flows
-    Callback::Callback<BasicSuccessCallback> mSuccess;
-    Callback::Callback<BasicFailureCallback> mFailure;
+    template <typename ClusterObjectT, typename RequestObjectT>
+    CHIP_ERROR SendCommand(DeviceProxy * device, RequestObjectT request,
+                           CommandResponseSuccessCallback<typename RequestObjectT::ResponseType> successCb)
+    {
+        ClusterObjectT cluster;
+        cluster.Associate(device, 0);
+
+        ReturnErrorOnFailure(cluster.InvokeCommand(request, this, successCb, OnFailure));
+        return CHIP_NO_ERROR;
+    }
 
     static CHIP_ERROR ConvertFromNodeOperationalCertStatus(uint8_t err);
 
-    Callback::Callback<OperationalCredentialsClusterCertificateChainResponseCallback> mCertificateChainResponseCallback;
-    Callback::Callback<OperationalCredentialsClusterAttestationResponseCallback> mAttestationResponseCallback;
-    Callback::Callback<OperationalCredentialsClusterOpCSRResponseCallback> mOpCSRResponseCallback;
-    Callback::Callback<OperationalCredentialsClusterNOCResponseCallback> mNOCResponseCallback;
-    Callback::Callback<DefaultSuccessCallback> mRootCertResponseCallback;
-    Callback::Callback<DefaultFailureCallback> mOnCertificateChainFailureCallback;
-    Callback::Callback<DefaultFailureCallback> mOnAttestationFailureCallback;
-    Callback::Callback<DefaultFailureCallback> mOnCSRFailureCallback;
-    Callback::Callback<DefaultFailureCallback> mOnCertFailureCallback;
-    Callback::Callback<DefaultFailureCallback> mOnRootCertFailureCallback;
+    inline bool SessionEstablishmentErrorIf(CHIP_ERROR error)
+    {
+        return SessionEstablishmentErrorIf(!chip::ChipError::IsSuccess(error), error);
+    }
+
+    inline bool SessionEstablishmentErrorIf(bool expr, CHIP_ERROR error)
+    {
+        if (expr)
+        {
+            static constexpr const char * header = "[Session Establishment Failure]";
+            ChipLogError(Controller, "%s %s at %s:%d", header, ErrorStr(error), __FILE__, __LINE__);
+            OnSessionEstablishmentError(error);
+        }
+        return !expr;
+    }
+
+    // Callbacks trampolines
+    template <typename T>
+    static void OnSuccess(void * context, const T & data)
+    {
+        DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+        commissioner->CommissioningStageComplete(CHIP_NO_ERROR);
+    }
+
+    static void OnFailure(void * context, EmberAfStatus status)
+    {
+        DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+        commissioner->SessionEstablishmentErrorIf(true, static_cast<CHIP_ERROR>(status));
```

Comment:
This static_cast does not make sense.

I think the old code just used a generic failure code.  We can do that, or we can create a StatusIB from the EmberAfStatus (by converting it to an `InteractionModel::Status`; we have functions for that), and using the new "put a StatusIB into a CHIP_ERROR" bits we have now.

Reply:
Hopefully we are using `CHIP_ERROR` after your changes and the conversion to `Invoke` !